src/clients/UBAClient/UBAClientUtilities.ts

```diff
@@ -77,3 +83,296 @@ export async function getUBAFeeConfig(
+    );
+
+  // Filter out:
+  // - Fills that request refunds on a different chain.
+  // - Subsequent fills after an initial partial fill.
+  // - Slow fills.
+  const fills: UbaFlow[] = spokePoolClient.getFills().filter((fill: FillWithBlock) => {
+    const result =
+      fill.repaymentChainId === spokePoolClient.chainId &&
+      fill.fillAmount.eq(fill.totalFilledAmount) &&
+      fill.updatableRelayData.isSlowRelay === false &&
+      fill.blockNumber > (fromBlock as number) &&
+      fill.blockNumber < (toBlock as number);
+    return result;
+  });
+
+  const refundRequests: UbaFlow[] = spokePoolClient.getRefundRequests(fromBlock, toBlock).filter((refundRequest) => {
+    const result = refundRequestIsValid(chainId, chainIdIndices, spokePoolClients, hubPoolClient, refundRequest);
+    if (!result.valid && logger !== undefined) {
+      logger.info({
+        at: "UBAClient::getFlows",
+        message: `Excluding RefundRequest on chain ${chainId}`,
+        reason: result.reason,
+        refundRequest,
+      });
+    }
+
+    return result.valid;
+  });
+
+  // This is probably more expensive than we'd like... @todo: optimise.
+  const flows = sortEventsAscending(deposits.concat(fills).concat(refundRequests));
+
+  return flows;
+}
+
+function refundRequestIsValid(
+  chainId: number,
+  chainIdIndices: number[],
+  spokePoolClients: { [chainId: number]: SpokePoolClient },
+  hubPoolClient: HubPoolClient,
+  refundRequest: RefundRequestWithBlock
+): RequestValidReturnType {
+  const { relayer, amount, refundToken, depositId, originChainId, destinationChainId, realizedLpFeePct, fillBlock } =
+    refundRequest;
+
+  if (!chainIdIndices.includes(originChainId)) {
+    return { valid: false, reason: "Invalid originChainId" };
+  }
+  const originSpoke = spokePoolClients[originChainId];
+
+  if (!chainIdIndices.includes(destinationChainId) || destinationChainId === chainId) {
+    return { valid: false, reason: "Invalid destinationChainId" };
+  }
+  const destSpoke = spokePoolClients[destinationChainId];
+
+  if (fillBlock.lt(destSpoke.deploymentBlock) || fillBlock.gt(destSpoke.latestBlockNumber)) {
+    return {
+      valid: false,
+      reason:
+        `FillBlock (${fillBlock} out of SpokePool range` +
+        ` [${destSpoke.deploymentBlock}, ${destSpoke.latestBlockNumber}]`,
+    };
+  }
+
+  // Validate relayer and depositId.
+  const fill = destSpoke.getFillsForRelayer(relayer).find((fill) => {
+    // prettier-ignore
+    return (
+        fill.depositId === depositId
+        && fill.originChainId === originChainId
+        && fill.destinationChainId === destinationChainId
+        && fill.amount.eq(amount)
+        && fill.realizedLpFeePct.eq(realizedLpFeePct)
+        && fill.blockNumber === fillBlock.toNumber()
+      );
+  });
+  if (!isDefined(fill)) {
+    return { valid: false, reason: "Unable to find matching fill" };
+  }
+
+  const deposit = originSpoke.getDepositForFill(fill);
```

Comment:
See [here](https://github.com/across-protocol/sdk-v2/pull/266/files#r1246726140) for note on validating the fill

Reply:
Sounds good! Updated